### PR TITLE
HDDS-5961. [Ozone-Streaming] update the usage space of Containers in …

### PR DESCRIPTION
**Pull request by guohao-rosicky**
HDDS-5961. [Ozone-Streaming] update the usage space of Containers in the stream write

jira: https://issues.apache.org/jira/browse/HDDS-5961

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
hi @captainzmc  , can you help review this patch?

**Comment by guohao-rosicky** [comment]
The old StreamDataChannel implementation did not update the usedBytes in the Container metadata at write time.
There will be multiple Container types in the future, I put the implementation of DataChannel into the Container package.

**Comment by guohao-rosicky** [comment]
hi @szetszwo  , can you help review this patch?

**Comment by szetszwo** [comment]
@guohao-rosicky , I just have sync'ed the HDDS-4454 branch with master.  Could you update this pull request?

**Comment by guohao-rosicky** [comment]
@szetszwo  @captainzmc , I have completed the update.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  @Override
+  public void force(boolean metadata) throws IOException {
+    randomAccessFile.getChannel().force(metadata);
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    int writeBytes = randomAccessFile.getChannel().write(src);
+    metrics.incContainerBytesStats(ContainerProtos.Type.WriteChunk, writeBytes);
+    containerData.updateWriteStats(writeBytes, false);
```
Comment (captainzmc) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:77]: Now we had updae metrics.incContainerBytesStats and containerData.updateWriteStats. Can you also add some test case? Just make sure that future changes do not affect this logic.

**Review by captainzmc — COMMENTED** [review]
LGTM overall,  Failed UT may not be caused by this pr, it can be triggered again.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java [hunk 2]
```diff
@@ -501,6 +501,23 @@ private ContainerCommandResponseProto runCommand(
...
+                      "traceID={}", gid, requestProto.getCmdType(),
+              requestProto.getContainerID(), requestProto.getPipelineID(),
+              requestProto.getTraceID());
+    }
+    try {
+      runCommand(requestProto, context);  // stream init
+      return dispatcher.getStreamDataChannel(requestProto);
+    } catch (IOException e) {
+      throw new CompletionException("Failed to create data stream", e);
+    }
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java:518]: Do not try-catch and throws StorageContainerException in the getStreamDataChannel(..) declaration.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java [hunk 3]
```diff
@@ -175,6 +176,16 @@ public VolumeChoosingPolicy getVolumeChoosingPolicyForTesting() {
     return volumeChoosingPolicy;
   }
 
+  public StateMachine.DataChannel getStreamDataChannel(
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:179]: Add `@Override`.

**Review by szetszwo — CHANGES_REQUESTED** [review]
@guohao-rosicky , thanks a lot for working on this.  Some comments inlined; see also https://issues.apache.org/jira/secure/attachment/13036136/2833_review.patch .

BTW, we use 4 spaces for continuation indent in Ozone.  Please change the continuation indent to 4 spaces (instead of 8).  Thanks.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Override
+  public void force(boolean metadata) throws IOException {
+    randomAccessFile.getChannel().force(metadata);
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    int writeBytes = randomAccessFile.getChannel().write(src);
+    metrics.incContainerBytesStats(ContainerProtos.Type.WriteChunk, writeBytes);
```
[the remark this answers is not in the document]
Comment (guohao-rosicky) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:76]: @szetszwo is container type.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java (hunk 1 above)
Comment (guohao-rosicky) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/KeyValueStreamDataChannel.java:69]: @captainzmc ,Has been modified.

**Comment by guohao-rosicky** [comment]
Has been modified, @szetszwo  @captainzmc .Please Take A Look.

**Review by szetszwo — APPROVED** [review]
+1 the latest change looks good.

**Pull request merged**